### PR TITLE
[POAE7-961] fix null pointer issue when offheap enabled - part II

### PR DESCRIPTION
**Pull request by yma11**
## What changes were proposed in this pull request?
bug fix for  null pointer issue when offheap enabled

## How was this patch tested?

PT

### CODE REVIEW

**Comment by yma11** [comment]
@winningsix Please help merge this PR.

File: src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java [hunk 1]
```diff
@@ -184,7 +184,7 @@ public void updateLongArray(LongArray sortedArray, int numRecords, int position)
             // update recordPointer in this array
             long originalRecordPointer = sortedArray.get(position);
             MemoryBlock page = taskMemoryManager.getOriginalPage(originalRecordPointer);
-            long baseOffset = page == null? 0: page.getBaseOffset();
+            long baseOffset = page.getBaseOffset();
```
Comment (winningsix) [inline src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:187]: Should be 
```
(page == null)? 0: page.getBaseOffset()
```
?

File: src/main/java/org/apache/spark/memory/TaskMemoryManager.java [hunk 2]
```diff
@@ -544,6 +544,12 @@ public MemoryBlock getOriginalPage(long pagePlusOffsetAddress) {
       assert (page != null);
       assert (page.getBaseObject() != null);
       return page;
+    } else if (tungstenMemoryMode == MemoryMode.OFF_HEAP) {
```
Comment (winningsix) [inline src/main/java/org/apache/spark/memory/TaskMemoryManager.java]: Only difference between on-heap and off-heap is one more check baseobj in L545?

File: src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java (hunk 1 above)
Comment (yma11) [reply src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:187]: No. for offheap page, the pageObject is null but not page itself is null.

File: src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 2 above)
Comment (yma11) [reply src/main/java/org/apache/spark/memory/TaskMemoryManager.java]: Indeed it's yes. Maybe we remove this assert (page.getBaseObject() != null) and unify the paths together?

File: src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java (hunk 1 above)
Comment (winningsix) [reply src/main/java/org/apache/spark/util/collection/unsafe/sort/PMemWriter.java:187]: ok

File: src/main/java/org/apache/spark/memory/TaskMemoryManager.java (hunk 2 above)
Comment (winningsix) [reply src/main/java/org/apache/spark/memory/TaskMemoryManager.java]: yes. we should

Comment (yma11) [reply src/main/java/org/apache/spark/memory/TaskMemoryManager.java]: done.

**Pull request merged**